### ISSUE TITLE
Multiple CSP headers appear to be misparsed

### ISSUE THREAD

**Issue by oschwald** [mozilla/http-observatory#396]
When a site has multiple CSP headers, it appears that the HTTP observatory [misparses it](https://observatory.mozilla.org/analyze/www.maxmind.com). My understanding of the specification is that if there are multiple CSPs, they should both be applied separately, essentially treating the requirements as a logical AND.

**Comment by floatingatoll** [comment]
CSP doesn't support merging of multiple CSP headers when one or more of
them use `report-uri`, or at least it's "undefined behavior". This isn't
well-reported by the Observatory, but cspvalidator.org says

> CSP merging error: Cannot merge policies if either policy contains a
report-uri directive.

**Comment by oschwald** [comment]
Ahh, that is interesting. I hadn't thought of that being the reason.

Looking at the [CSP 2 specification](https://www.w3.org/TR/CSP2/#implementation-considerations), it appears that the `report-uri` only applies to the header value it is applied to:

> Server administrators MAY wish to send multiple policies if different reporting options are desired for subsets of an overall policy. For instance, the following headers:
>
> `Content-Security-Policy: frame-ancestors https://example.com/ `
>
> `Content-Security-Policy: default-src https:; report-uri https://example.com/`
>
> would send violation reports for http resources, but would not send violation reports for frame-ancestors violations. Note also that combining them via ',' into the single header
>
> `Content-Security-Policy: frame-ancestors https://example.com/, default-src https:; report-uri https://example.com`
>
> would have the same effect, as the comma splits the header during parsing.



I haven't examined the specification closely, but I am not finding anything that suggests that the usage is incorrect or undefined.

**Comment by oschwald** [comment]
Reading your response more closely, perhaps the issue is that the directive for things like `script-src` are being merged as opposed to the example from the spec where the directives are distinct. I'll take a closer look when I have a few minutes.

**Comment by floatingatoll** [comment]
Yeah, I couldn't find a great deal of explanation around this, but I did
see enough warnings about it that — essentially — while you _may_ be able
to do so, you _won't_ find that it works well or reliably. You could try
running the Observatory in local scanner mode —
https://github.com/mozilla/http-observatory/README.md — and trace the code
to find out what specific issue is being logged, but ultimately I think
it's going to boil down to "a python CSP parsing module has decided that
you can't do that", and I suspect this isn't a scenario where the
Observatory will be altered to accept this usage.

If you try merging all your CSP headers into a single header, it at least
passes cspvalidator.org, but it spits out this warning three times, so
hopefully all three report-uri are the same, assuming they're parsed
successfully by the useragent:

1:757: Policy contains more than one report-uri directive. All but the
first instance will be ignored.

As a final note, I see that you have script-src: unsafe-inline, so that
"unable to parse CSP" (-25) will turn into
"csp-implemented-with-unsafe-inline" (-20) if/when you consolidate CSP
headers or remove all uses of report-uri:

> Content Security Policy (CSP) implemented unsafely. This includes
`\'unsafe-inline\'` or `data:` inside script-src, overly broad sources such
as `https:` inside `object-src` or `script-src`, or not restricting the
sources for `object-src` or `script-src`. | -20

**Comment by oschwald** [comment]
Unfortunately, it is not possible to merge the two headers into a single header and retain the same meaning. Each header is a further restriction on top of the previous one. As such, you can for instance, have both a URL restriction and a `nonce` be required. For that reason, the `unsafe-inline` you mention isn't as much of an issue as a `nonce` is required by the other policy.

**Comment by oschwald** [comment]
There was some additional discussion on the repo for cspvalidator.org in https://github.com/shapesecurity/salvation/issues/228.

**Comment by floatingatoll** [comment]
I think that I can safely say that your knowledge of and use of CSP here
far exceeds that of, it appears, every automated grading system you’ve
encountered to date. This makes my day :)

If you chain the multiple consecutive headers together with semicolons (not
commas, not reordering at all) into a single header, they _aren’t_ parsed
as if they were separated by Header: lines?

That seems wrong, at a CSP spec level. But maybe that’s just how it is.

(I want to be sure I’m understanding that right, because they seem to be
talking about reordering them in 228 and that’s not what I’m wondering
about.)

**Comment by mozdevcontrib** [comment]
`Content-Security-Policy` as well as `Feature-Policy` are quickly evolving, An example for CSP 
is - `style-src-elem`  ,  `style-src-attr` &  their script equivalents. For `Feature-Policy` Mozilla & Google have different terms to define what "features" to lock down. Mozilla uses `xr` and Google uses `vr` to define virtual reality. Technically, Mozilla is correct, while Google uses virtual reality's initials. And thus far, Google wins out, unfortunately. Personally, I use terms from both Mozilla & Google as I'm forward thinking. And one day, Mozilla's terminology _will at least be on par_ with Google's. See here: https://www.chromestatus.com/feature/5141352765456384 & here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src , as well as the parent pages to really see the conflict that's ongoing. **_Go Mozilla_**!

**Comment by april** [comment]
A lot of this is a result of the way that Python/requests handle HTTP requests that have duplicate headers. When they encounter them, they combine them with a comma and (afaik) provide no indication that this happened.

This makes it difficult/impossible for me to know if there were duplicate headers or if there was actually a comma there.

**Comment by oschwald** [comment]
The comma should split the header when parsing and it should be treated as two separate policies. The spec treats a single header with comma separated values the same as multiple headers.
